### DISCUSSION TITLE
Support for custom components

### DISCUSSION THREAD

**Discussion by paulmbw** [fabricjs/fabric.js#7465] [Q&A]
Hi there, 

Tl;dr: I would like the canvas to support rending custom React components, for example:

```
export default function Editor() {
  return (
    <div>
      <h2>Editor Component</h2>
      <Editor />
    </div>
  );
}
```

This could be a WYSIWYG editor for example. The only properties this component will require is to be able to drag the component across the canvas. This is possible?

**Comment by asturur** [comment]
Hi @paulmbw is unclear what do you mean.

**Comment by paulmbw** [comment]
Apologies for not being clearer. It appears this library only supports basic shapes such as `Rect`, `Text` and `Image`. However, I need the canvas to support a React component of my choosing, for example a React component that renders a text editor. This component will also need similar properties to that of `Rect` and so on, for instance, being able to drag the component and resizing. Would this be possible?

**Reply by paulmbw** [reply]
@asturur do you have any thoughts on this please?

**Reply by andrewzolotukhin** [reply]
Fabric renders on Canvas. You can create a custom component,  but it should inherit base Fabric Object class. I think your question is irrelevant to this library.

**Reply by paulmbw** [reply]
Ah I see, makes sense. Thank you for clarifying. God bless.

**Reply by asturur** [reply]
Yes as @andrewzolotukhin said, this is a graphic render tool. You can render images, shapes, complex shapes, text, gradients. Things you would see in photshop more than in word for example.
Is a rendering library with many similarities and funcionalities shared with a content editor

**Reply by paulmbw** [reply]
Okay, great. And just to clarify once more, if I wanted to render objects that are outside of the fabric domain (for example, create a react component that renders my own text editor for example) I would need to extend the Object class? Do you happen to have an example of how this is done? Most other examples on GitHub are extending from Rect or Image, but I would like to extend from the Object class and render my own react component

**Reply by asturur** [reply]
you won't be able to render a react component, if not creating a subclass that will drag around a dom element over the canvas, as an extra layer, positioned absolutely, and then binding this dom element both to the object and to react.
It does not look like a great idea in general, i would consider alternatives.

This render will happen on the browser but won't be captured on the JPG/PNG that you may want to create, nor in the SVG or can be serialized without react context.

**Reply by paulmbw** [reply]
Right, makes sense. I've had a look at alternatives such as PixiJs which is what Miro appear to be using. For some context, we are building a collaborative whiteboard where you can add text and images, but we have our own custom components already built that need to be rendered on a canvas. Maybe a canvas isn't that right tool? Not sure, continuing to investigate. Thanks a lot for your feedback and responses, I really appreciate the time you've given to respond.

Take care and God bless :)

**Reply by asturur** [reply]
PixiJS does not make much difference ( well is a faster library )
You should probably decide what to display on the whiteboard. DOM or CANVAS. a mixed approach is just extra work and asking for troubles.

Canvas can draw text, and DOM can draw primitives just fine with SVGs.

**Reply by paulmbw** [reply]
Makes sense, I think we will have to go with rendering on the DOM for now as a first version. My understanding is canvas applications perform much better, especially when it comes to collaborating with other users.

**Comment by rkurbatov** [comment]
It was pretty hard to marry React and Fabric.js, we put a lot of effort in it.

We had a special react-context hook keeping the fabric canvas instance and method to set it on init among some settings, required for our editor, but then decided to move as much as possible to the canvas object itself to prevent deep nested property passing.

For example, our canvas was extended with reduxDispatch method that allows dispatching redux actions in various handlers. That allows us to have history in redux state. We have a handler for selection that dispatches the action like 'objectSelected' putting some flag in the store. If the flag is present in the store we get it from the selector and draw something like context menu as a react component in a div, covering the canvas. So we have real canvas layer => fabirc.js canvas layer for auxiliary drawings => react div for context menu and some other components related to UI on the very top. You can work on that additional layer and keep there some high level UI.

The canvas is put into the context on the top editor component and can be used anywhere you need it. So two way binding is like so: dispatching events to redux in canvas handlers, putting some value in the redux store, selecting that events in React components, calling canvas handlers in react components.

**Reply by asturur** [reply]
Is not that is hard or easy, is that you have to keep them in sync and avoid loops of updates.
Everyone will land on a specific solution that match his app and needs, that is why apart a bunch of small hooks i never suggested anything as the way to go.

**Comment by andrewzolotukhin** [comment]
There is no problem to use Fabric inside a React app. For example we did in cleverbrush.com.

**Reply by paulmbw** [reply]
And you were able to render react components on the canvas?

**Reply by rkurbatov** [reply]
Surely not! Canvas is a canvas, it allows drawing bitmaps and primitives. React components are "drawn" as DOM nodes. Fabric.js just provides the separate layer of high-level abstraction and API over primitive canvas.

**Reply by ShaMan123** [reply]
#6208
